Add `--extra-inputs` to `containers-add`

nobodyinperson opened the pull request:
This PR adds the `datalad containers-add --extra-inputs` option to specify additional files that should be included in the run-record and thus auto-fetched upon `datalad containers-run`.

Furthermore the `--call-fmt` now also has a `{img_dir}` placeholder for easier specification of files next to the image.

I added a test for `containers-add` asserting that the correct values land in `.datalad/config`. There is currently no test for `containers-run` as I couldn't get this to work in CI. I tried a shell-script as 'container' and sourced files as 'overlays', works on my machine, weird errors in CI. 

Full test script `test/run.sh` for reference, a comparable `pytest` is shipped with this PR:

```bash
#!/bin/sh
cd $(dirname "$(readlink -f "$0")")
(chmod -R +w ds;rm -rf ds) 2>/dev/null
set -x
datalad create ds
cd ds
mkdir containers
sudo apptainer build containers/alpine.sif docker://alpine
sudo apptainer overlay create --size 64 containers/overlay.img
sudo chown $USER:$USER containers/overlay.img
sudo apptainer overlay create --size 64 containers/overlay2.img
sudo chown $USER:$USER containers/overlay2.img
datalad $@ containers-add alpine -i containers/alpine.sif
datalad $@ containers-add alpine-with-overlay \
    --call-fmt 'apptainer exec --overlay {img_dir}/overlay.img {img} {cmd}' \
    --extra-inputs containers/overlay.img \
    -i containers/alpine.sif
datalad $@ containers-add alpine-with-two-overlays \
    --call-fmt 'apptainer exec --overlay {img_dir}/overlay.img --overlay {img_dir}/overlay2.img:ro {img} {cmd}' \
    --extra-inputs containers/overlay.img containers/overlay2.img \
    -i containers/alpine.sif
datalad save
echo 1 > file.txt
datalad save
cat .datalad/config
datalad $@ containers-run -n alpine -o alpine-uname-a.txt 'uname -a > {outputs[0]}'
cat alpine-uname-a.txt
datalad $@ containers-run -n alpine-with-overlay -o alpine-with-overlay-uname-a.txt 'uname -a > {outputs[0]}'
cat alpine-with-overlay-uname-a.txt
datalad $@ containers-run -n alpine-with-two-overlays -o alpine-with-two-overlays-uname-a.txt 'uname -a > {outputs[0]}'
cat alpine-with-two-overlays-uname-a.txt
```

closes #189

bpoldrack commented:
Thanks for approaching that, @nobodyinperson!

WRT to the CI failure: The new parameter needs to also become part of the actual signature:
https://github.com/datalad/datalad-container/pull/190/files#diff-74ae0e7052ad9775f337a738a2a3915e46a23f555f2acbf4c24c72f6ebbab4caR182

nobodyinperson commented:
> Thanks for approaching that, @nobodyinperson!
> 
> WRT to the CI failure: The new parameter needs to also become part of the actual signature: https://github.com/datalad/datalad-container/pull/190/files#diff-74ae0e7052ad9775f337a738a2a3915e46a23f555f2acbf4c24c72f6ebbab4caR182

Yep, working on it!

bpoldrack commented:
Thank you very much, @nobodyinperson . I'm looking into the test issues to finish this.

yarikoptic commented on datalad_container/containers_run.py:
```diff
@@ -131,14 +132,34 @@ def __call__(cmd, container_name=None, dataset=None,
+            ).format(container.get("extra-inputs"),type(extra_inputs),extra_inputs))
+        extra_inputs = list(map(str, 
+            [extra_inputs] if isinstance(extra_inputs, str) else extra_inputs
+        ))
```
if we just store multiple values in the config in multiple lines, then smth like `list(config.get("extra-inputs", [], get_all=True))` should be all what is needed since by default values are `str` already and you might just get a tuple not a list may be

yarikoptic reviewed (changes requested):
brief comments on how to harmonize with other interfaces etc

nobodyinperson commented:
@yarikoptic I implemented all of your suggestions. I also made `img_dirpath` and `img_dspath` available in `--extra-input` and added a check in `containers-add` for that (for `--call-fmt` invalid placeholers are silently ignored until `containers-run`, might want to add a similar check there as well).

I tried again to make a test for `containers-run` with `extra-input` but I couldn't get it work. Please one of you experts who are more familiar with the code base step in as I am hitting a dead end here and just burning time.

yarikoptic commented:
just to make sure -- do you mean TODO at https://github.com/datalad/datalad-container/pull/190/files#diff-c9a6a0cde74f0c35ded765fc396d537f9485fe47645c4e00698ab951138042d6R218 ?

nobodyinperson commented:
> just to make sure -- do you mean TODO at https://github.com/datalad/datalad-container/pull/190/files#diff-c9a6a0cde74f0c35ded765fc396d537f9485fe47645c4e00698ab951138042d6R218 ?

Yes, in general a test for `containers-run`. The above linked comments are not enough to test the new `--extra-input` feature.

yarikoptic commented on datalad_container/containers_add.py:
```diff
@@ -321,6 +339,25 @@ def __call__(name, url=None, dataset=None, call_fmt=None, image=None,
+                yield get_status_dict(
+                    action="containers_add", ds=ds, logger=lgr,
+                    status="error",
+                    message=("--extra-input %r contains unknown placeholder %s. "
```
@mih @adswa - I forgot but remember we had some conversation: do we have some syntax sugaring to use in `message` of records to differentiate CLI `--option` from Python API function `option`?

adswa replied:
not that I'm aware of

yarikoptic replied:
ok, thanks!

yarikoptic commented:
@nobodyinperson I sent a PR for your fork with suggestive changes/fixup to the test: https://github.com/nobodyinperson/datalad-container/pull/1

nobodyinperson commented:
Thanks @yarikoptic. How is the run-record accessed? The `´ds.repo` apparently can't access the `git log`.

yarikoptic commented:
probably look at `rerun` for somewhat ultimate referral on how "accessed": https://github.com/datalad/datalad/blob/HEAD/datalad/local/rerun.py#L282

yarikoptic commented on datalad_container/containers_run.py:
```diff
@@ -131,14 +132,41 @@ def __call__(cmd, container_name=None, dataset=None,
+                        ", ".join(xi_kwargs)))
+                return
+
+        lgr.debug("extra_inputs = {!r}".format(extra_inputs))
```
```suggestion
        lgr.debug("extra_inputs = %r", extra_inputs)
```
no need to do formatting unless really at debug level. I know that savings are negligible but in general it is a preferred way (also to avoid surprises if `extra_inputs` has some `%` in them and then someone would add here more to the line, and some argument to interpolate and it would all break)

yarikoptic commented on datalad_container/containers_run.py:
```diff
@@ -131,14 +132,41 @@ def __call__(cmd, container_name=None, dataset=None,
                     dataset=dataset or (ds if ds.path == pwd else None),
                     inputs=inputs,
-                    extra_inputs=[image_path],
+                    extra_inputs=[image_path] + (extra_inputs or []),
```
if I got indentation right, we always guaranteed to have `extra_inputs`  initialized to `[]` so may be just 

```suggestion
                    extra_inputs=[image_path] + extra_inputs,
```

yarikoptic commented on datalad_container/containers_run.py:
```diff
@@ -131,14 +132,41 @@ def __call__(cmd, container_name=None, dataset=None,
+        # TODO: Can't use EnsureListOf(str) yet as it handles strings as iterables.
+        # See this PR: https://github.com/datalad/datalad/pull/7267
+        # get_all=True still returns a string for a single value
+        extra_inputs_ = [extra_inputs_] if isinstance(extra_inputs_, str) else extra_inputs_
```
in principle, we have `ensure_list` which does that  https://github.com/datalad/datalad/blob/maint/datalad/utils.py#L734  so could just wrap around above `container.get` and avoid needing extra `extra_inputs_`

yarikoptic commented on datalad_container/containers_run.py:
```diff
@@ -131,14 +132,41 @@ def __call__(cmd, container_name=None, dataset=None,
             # just prepend and pray
             cmd = container['path'] + ' ' + cmd
 
+        extra_inputs_ = container.get("extra-input",[])
```
we aren't really flake'ing etc , but we do love to separate them out:

```suggestion
        extra_inputs_ = container.get("extra-input", [])
```

yarikoptic commented on datalad_container/tests/test_run.py:
```diff
@@ -191,6 +192,50 @@ def test_custom_call_fmt(path=None, local_file=None):
+            "sub/overlays/overlay3.img",
+        ]
+    ) == set(runinfo.get("extra_inputs", set()))
+
```
```suggestion


```

two empty lines typically to separate out.  

also please remove debugging commented out `print` statements if no loner needed for your development . We generally don't have such ones, so I don't think worth adding them here either

yarikoptic reviewed (changes requested):
Looks good to me . Just nit picking to make a little more consistent with the rest of inconsistencies we have ;)

nobodyinperson commented:
@yarikoptic I addressed all your remarks and rebased the branch onto `master` to get rid of the conflict around `from datalad.interface.{utils,base} import eval_results`.

Concerning the code style discussions I'd recommend using [black](https://github.com/psf/black) as code formatter. Takes the discussion out of the formatting, it's one de-facto standard for Python code style.

yarikoptic commented on datalad_container/tests/test_run.py:
```diff
@@ -191,6 +192,51 @@ def test_custom_call_fmt(path=None, local_file=None):
+    )
+    ds.save(recursive=True)  # record the entire tree of files etc
+    ds.containers_run("XXX", container_name="sub/mycontainer")
+    # print(ds.repo.call_git(["ls-files"]))
+    ok_file_has_content(
+        os.path.join(ds.repo.path, "out.log"),
+        "image=sub/containers/container.img",
+        re_=True,
+    )
+    # print(ds.repo.call_git(["log"]))
+    commit_msg = ds.repo.call_git(["show", "--format=%B"])
+    # print(f"{commit_msg = }")
+    cmd, runinfo = get_run_info(ds, commit_msg)
+    # print(f"{runinfo = }")
```
```suggestion
    ok_file_has_content(
        os.path.join(ds.repo.path, "out.log"),
        "image=sub/containers/container.img",
        re_=True,
    )
    commit_msg = ds.repo.call_git(["show", "--format=%B"])
    cmd, runinfo = get_run_info(ds, commit_msg)
```

yarikoptic replied:
I will just "do it" to not delay rerunning CI.

yarikoptic commented on changelog.d/pr-190.md:
```diff
@@ -0,0 +1,3 @@
+### 🚀 Enhancements and New Features
+
+- Add `--extra-inputs` to `containers-add`.  Fixes [#189](https://github.com/datalad/datalad-container/issues/189) via [PR #190](https://github.com/datalad/datalad-container/pull/190) (by [@nobodyinperson](https://github.com/nobodyinperson))
```
feel welcome to expand here on how/when to use this feature, e.g. close to your use case.

yarikoptic commented on datalad_container/containers_run.py:
```diff
@@ -131,14 +133,36 @@ def __call__(cmd, container_name=None, dataset=None,
+                xi_kwargs = dict(
+                    img_dspath=image_dspath,
+                    img_dirpath=op.dirname(image_path) or ".",
+                )
```
codeclimate btw warned about code duplication here. I think it is ok, although indeed duplicates with cmd_kwargs definition above so we could have some `common_kwargs` to be this structure and defined before any use and then only enhanced for `cmd_kwargs`.  But really might be not worth the hassle this far, just wanted to leave a note that I am ok to suppress codeclimate for this.

yarikoptic commented:
pushed potential fix for appveyor failure to configure singularity. If doesn't work out -- can troubleshoot in a separate PR
